add k8s docs for getting started, K8s Manifest and Helm (#179)
opea-project/docs: examples/ChatQnA/deploy/k8s_helm.md

# Multi-node on-prem deployment with TGI on Xeon Scalable processors on a K8s cluster using Helm

> This deployment section covers multi-node on-prem deployment of the ChatQnA example with OPEA comps to deploy using the TGI service. There are several slice-n-dice ways to enable RAG with vectordb and LLM models, but here we will be covering one option of doing it for convenience: we will be showcasing how to build an e2e chatQnA with Redis VectorDB and neural-chat-7b-v3-3 model, deployed on a Kubernetes cluster using Helm. For more information on how to setup a Xeon based Kubernetes cluster along with the development pre-requisites, follow the instructions here [Kubernetes Cluster and Development Environment](./k8s_getting_started.md#kubernetes-cluster-and-development-environment). For a quick introduction on Helm Charts, visit the helm section in [Getting Started with Kubernetes for ChatQnA](./k8s_getting_started.md).

Author: @arun-gupta As discussed on teams -- this has also been tested on multi-node K8s cluster.

```bash
git clone https://github.com/opea-project/GenAIInfra.git
```
Checkout the release tag
```
cd GenAIInfra/helm-charts/
git checkout tags/v1.1
```
### HF Token
The example can utilize model weights from HuggingFace and langchain.

Setup your [HuggingFace](https://huggingface.co/) account and generate
[user access token](https://huggingface.co/docs/transformers.js/en/guides/private#step-1-generating-a-user-access-token).

Setup the HuggingFace token
```
export HF_TOKEN="Your_Huggingface_API_Token"
```

### Proxy Settings

If you are behind a corporate VPN, proxy settings must be added for services requiring internet access, such as the LLM microservice, embedding service, reranking service, and other backend services.
Proxy can be set in the `values.yaml`.
Open the `values.yaml` file using an editor
```bash
vi chatqna/values.yaml
```
Update the following section and save file:
```yaml
# chatqna/values.yaml
global:
  http_proxy: "http://your-proxy-address:port"
  https_proxy: "http://your-proxy-address:port"
  no_proxy: "localhost,127.0.0.1,localaddress,.localdomain.com"
```
## Use Case Setup

> The `GenAIInfra` repository utilizes a structured Helm chart approach, comprising a primary `Charts.yaml` and individual sub-charts for components like the LLM Service, Embedding Service, and Reranking Service. Each sub-chart includes its own `values.yaml` file, enabling specific configurations such as Docker image sources and deployment parameters. This modular design facilitates flexible, scalable deployment and easy management of the GenAI application suite within Kubernetes environments. For detailed configurations and common components, visit the [GenAIInfra common components directory](https://github.com/opea-project/GenAIInfra/tree/main/helm-charts/common).

Reviewer: "Docker image sources" => "container image name/version"

This use case employs a tailored combination of Helm charts and `values.yaml` configurations to deploy the following components and tools:
|use case components | Tools |   Model     | Service Type |
|----------------     |--------------|-----------------------------|-------|
|Data Prep            |  LangChain   | NA                       |OPEA Microservice |
|VectorDB             |  Redis       | NA                       |Open source service|
|Embedding            |   TEI        | BAAI/bge-base-en-v1.5    |OPEA Microservice |
|Reranking            |   TEI        | BAAI/bge-reranker-base   | OPEA Microservice |
|LLM                  |   TGI        |Intel/neural-chat-7b-v3-3 |OPEA Microservice |
|UI                   |              | NA                       | Gateway Service |

Tools and models mentioned in the table are configurable either through the
environment variable or `values.yaml` 

Set a new [namespace](#create-and-set-namespace) and switch to it if needed

> To enable UI, uncomment the lines `56-62` in `GenAIInfra/helm-charts/chatqna/values.yaml`:

Author: @arun-gupta The lines have now changed to 59-63 that need to be uncommented. I'll remove the pointer to the liner numbers and say uncomment the following.
Could you tell me which space needs to be deleted?

```bash
chatqna-ui:
   image:
     repository: "opea/chatqna-ui"
     tag: "latest"
   containerPort: "5173"
```

Next, we will update the dependencies for all Helm charts in the specified directory and ensure the `chatqna` Helm chart is ready for deployment by updating its dependencies as defined in the `Chart.yaml` file.

```bash
# all Helm charts in the specified directory have their 
# dependencies up-to-date, facilitating consistent deployments.
./update_dependency.sh

# "chatqna" here refers to the directory name that contains the Helm 
# chart for the ChatQnA application 
helm dependency update chatqna
```

Set the necessary environment variables to setup the use case
```bash

> export MODELDIR="/mnt/opea-models"  #export MODELDIR="null" if you don't want to cache the model.  

Reviewer 2: This caused an error:

```
Events:
  Type     Reason       Age                 From               Message
  ----     ------       ----                ----               -------
  Normal   Scheduled    10m                 default-scheduler  Successfully assigned default/chatqna-tei-645548d7f-7crjj to minikube
  Warning  FailedMount  17s (x13 over 10m)  kubelet            MountVolume.SetUp failed for volume "model-volume" : hostPath type check failed: /mnt/opea-models is not a directory
```

Using this command `export MODELDIR="null"` solved the issue.

Reviewer 3: I would not recommend to set global.modelUseHostPath=${MODELDIR} in `helm install` in multi-node environment. This options is meant to share the pre downloaded model data files in a single node environment for a quick setup to test. To use this in a multiple nodes environment , you need to go to every K8S worker node to make sure there is a `${MODELDIR}` directory exists and writable.  Just don't set global.modelUseHostPath and leave it as the default value(empty string) which will download the models without sharing.

Another option to share the model files is to set global.modelUsePVC to use K8S persistent volume to share the model data files. In order to use that, users need to do some preparation work out-of-band:

1. Setup the K8S storage class, persistent volumes.
2. Create a K8S persistent volume claim which can be shared by multi pods.
3. pass the created k8s persistent volume claim to global.modelUsePVC

Reviewer: I would not recommend setting it empty either, because not sharing the data will much more easily cause node to run out disk space, meaning that k8s evicts pods from it, and things in general break.

Author: @eero-t @lianhao 
Could we leave it empty and point to the following link to set up pvc?

https://github.com/opea-project/GenAIInfra/blob/56674fcc5ad76cdcf9f17377cef4f5ff8e3b90a4/helm-charts/README.md#using-persistent-volume:~:text=their%20usage%20metrics.-,Using%20Persistent%20Volume,-It%27s%20common%20to

Reviewer: Linked instruction is fairly OK, but it assumes user knows something about setting up NFS:
* security implication of non-authenticated access
* need to replace IP address with the correct one for the used NFS server

Some knowledge of how PVCs work may also be good in case one needs to debug why pods are in `Pending` state (is the problem in NFS settings, PVC not in same namespace with pod using it, there being too many PVCs for given PV etc).

```

**OUTPUT:**
```bash
NAME: chatqna
LAST DEPLOYED: Thu Sep  5 13:40:20 2024
NAMESPACE: chatqa
STATUS: deployed
REVISION: 1
```
It takes a few minutes for all the microservices to be up and running. Go to the next section which is [Validate Microservices](#validate-microservices) to verify that the deployment is successful.

### Validate microservice
#### Check the pod status
To check if all the pods have started, run:

```bash
kubectl get pods
``` 
You should expect a similar output as below:
```
NAME                                      READY   STATUS    RESTARTS      AGE
chatqna-chatqna-ui-77dbdfc949-6dtms        1/1     Running   0          5m7s
chatqna-data-prep-798f59f447-4frqt         1/1     Running   0          5m7s
chatqna-df57cc766-t6lkg                    1/1     Running   0          5m7s
chatqna-nginx-5dd47bfc7d-54x96             1/1     Running   0          5m7s
chatqna-redis-vector-db-7f489b6bb6-mvzbw   1/1     Running   0          5m7s
chatqna-retriever-usvc-6695979d67-z5jgx    1/1     Running   0          5m7s
chatqna-tei-769dc796c-gh5vx                1/1     Running   0          5m7s
chatqna-teirerank-54f58c596c-76xqz         1/1     Running   0          5m7s

> chatqna-tgi-7b5556d46d-pnzph               1/1     Running   0          5m7s

Reviewer 2: Could not get all pods to run, filed https://github.com/opea-project/GenAIExamples/issues/1202

Author: @arun-gupta I was thinking of adding separate resource requirements for - on-prem (baremeral) servers and cloud instances
Does that make sense?

kubernetes                ClusterIP   10.92.0.100      <none>        443/TCP             11d
 ```
   To begin port forwarding, which maps a service's port from the cluster to local host for testing, use:
 ```bash
    kubectl port-forward svc/[service-name] [local-port]:[service-port]
   ```
   Replace `[service-name]`, `[local-port]`, and `[service-port]` with the appropriate values from your services list (as shown in the output given by `kubectl get svc`). This setup enables interaction with the microservice directly from the local machine. In another terminal, use `curl` commands to test the functionality and response of the service.

Use `ctrl+c` to end the port-forwarding to test other services.

### MegaService Before RAG Dataprep

Use the following command to forward traffic from your local machine to the service running in the Kubernetes cluster:
```bash
kubectl port-forward svc/chatqna 8888:8888
```
Follow the below steps in a different terminal.

```
curl http://localhost:8888/v1/chatqna -H "Content-Type: application/json" -d '{
     "model": "Intel/neural-chat-7b-v3-3",
     "messages": "What is the revenue of Nike in 2023?"

>      }'

Reviewer: Query specified here seems outdated based on answer?

Reviewer 2: Getting the following error when running this query:

```
ubuntu@ip-172-31-7-63:~/.minikube$ curl http://localhost:8888/v1/chatqna -H "Content-Type: application/json" -d '{
     "model": "Intel/neural-chat-7b-v3-3",
     "messages": "What is the revenue of Nike in 2023?"
     }'
curl: (7) Failed to connect to localhost port 8888 after 0 ms: Couldn't connect to server
```

Author: @eero-t Fixed the query

```
Here is the output for your reference:
```bash

> data: b' O', data: b'PE', data: b'A', data: b' stands', data: b' for', data: b' Organization', data: b' of', data: b' Public', data: b' Em', data: b'ploy', data: b'ees', data: b' of', data: b' Alabama', data: b'.', data: b' It', data: b' is', data: b' a', data: b' labor', data: b' union', data: b' representing', data: b' public', data: b' employees', data: b' in', data: b' the', data: b' state', data: b' of', data: b' Alabama', data: b',', data: b' working', data: b' to', data: b' protect', data: b' their', data: b' rights', data: b' and', data: b' interests', data: b'.', data: b'', data: b'', data: [DONE]

Reviewer: It does not look like this.  Each `data:` output is on its own line, with extra empty lines between.

Author: @eero-t 
These take up a lot of space, so I decided to wrap in a paragraph
The user can compare the two outputs and conclude if correct.

Reviewer: I do not think it needs to show whole response.  I would suggest showing just few lines from the start and end of the output, with `...` between those indicating there's more.

```
which is essentially the following sentence:
```

> OPEA stands for Organization of Public Employees of Alabama. It is a labor union representing public employees in the state of Alabama, working to protect their rights and interests.

Reviewer: "Revenue of Nike" is "Organization of Public Employees of Alabama"?

Use the following command to forward traffic from your local machine to the service running in the Kubernetes cluster:
```bash
kubectl port-forward svc/chatqna-data-prep 6007:6007
```
Follow the below steps in a different terminal.

If you want to add/update the default knowledge base, you can use the following
commands. The dataprep microservice extracts the texts from variety of data
sources, chunks the data, embeds each chunk using embedding microservice and
store the embedded vectors in the redis vector database.

this example leverages the OPEA document for its RAG based content. You can download the [OPEA document](https://opea-project.github.io/latest/_downloads/41c91aec1d47f20ca22350daa8c2cadc/what_is_opea.pdf) and upload it using the UI.

Local File `what_is_opea.pdf` Upload:

```
curl -X POST "http://localhost:6007/v1/dataprep" \
     -H "Content-Type: multipart/form-data" \
     -F "files=@./what_is_opea.pdf"
```

This command updates a knowledge base by uploading a local file for processing.
Update the file path according to your environment.

You should see the following output after successful execution:
```
{"status":200,"message":"Data preparation succeeded"}
```

> For advanced usage of the dataprep microservice refer [here](#dataprep-microservice-%28advanced%29)

Reviewer: Link is wrong (can be tested with "View file" functionality):
```suggestion
For advanced usage of the dataprep microservice refer [here](#dataprep-microservice-advanced).
```

### MegaService After RAG Dataprep

Use the following command to forward traffic from your local machine to the service running in the Kubernetes cluster:
```bash
kubectl port-forward svc/chatqna 8888:8888
```
Similarly, follow the below steps in a different terminal.

```
curl http://localhost:8888/v1/chatqna -H "Content-Type: application/json" -d '{
     "model": "Intel/neural-chat-7b-v3-3",
     "messages": "What is OPEA?"
     }'

> 

Reviewer: Please remove these redundant empty lines from all the command examples.  They do not look good.

```
OPEN Platform for Enterprise AI (Open Platform for Enterprise AI) is a framework that focuses on creating and evaluating open, multi-provider, robust, and composable generative AI (GenAI) solutions. It aims to facilitate the implementation of enterprise-grade composite GenAI solutions, particularly Retrieval Augmented Generative AI (RAG), by simplifying the integration of secure, performant, and cost-effective GenAI workflows into business systems.
```

### TEI Embedding Service
Use the following command to forward traffic from your local machine to the service running in the Kubernetes cluster:
```bash
kubectl port-forward svc/chatqna-tei 6006:80
```
Follow the below steps in a different terminal.

The TEI embedding service takes in a string as input, embeds the string into a
vector of a specific length determined by the embedding model and returns this
embedded vector.

```
curl http://localhost:6006/embed \
    -X POST \
    -d '{"inputs":"What is Deep Learning?"}' \
    -H 'Content-Type: application/json'
```

> In this example the embedding model used is "BAAI/bge-base-en-v1.5", which has a vector size of 768. So the output of the curl command is a embedded vector of

Reviewer: Please use correct formatting for commands and command lines:
```suggestion
In this example the embedding model used is "BAAI/bge-base-en-v1.5", which has a vector size of 768. So the output of the `curl` command is a embedded vector of
```

length 768.

### Retriever Microservice
Use the following command to forward traffic from your local machine to the service running in the Kubernetes cluster:
```bash
kubectl port-forward svc/chatqna-retriever-usvc 7000:7000
```
Follow the below steps in a different terminal.

To consume the retriever microservice, you need to generate a mock embedding
vector by Python script. The length of embedding vector is determined by the
embedding model. Here we use the
model EMBEDDING_MODEL_ID="BAAI/bge-base-en-v1.5", which vector size is 768.

Check the vector dimension of your embedding model and set
`your_embedding` dimension equal to it.

```

> export your_embedding=$(python3 -c "import random; embedding = [random.uniform(-1, 1) for _ in range(768)]; print(embedding)")

Reviewer: Both `export` and `embedding` var are redundant:
```suggestion
your_embedding=$(python3 -c "import random; print([random.uniform(-1, 1) for _ in range(768)])")
```

Author: @eero-t I believe it's more readable this way for the user, and the source for this is the dev team's readme.

curl http://localhost:7000/v1/retrieval \
  -X POST \
  -d "{\"text\":\"test\",\"embedding\":${your_embedding}}" \
  -H 'Content-Type: application/json'

```
The output of the retriever microservice comprises of the a unique id for the
request, initial query or the input to the retrieval microservice, a list of top
`n` retrieved documents relevant to the input query, and top_n where n refers to
the number of documents to be returned.

The output is retrieved text that relevant to the input data:
```

> {"id":"27210945c7c6c054fa7355bdd4cde818","retrieved_docs":[{"id":"0c1dd04b31ab87a5468d65f98e33a9f6","text":"Company: Nike. financial instruments are subject to master netting arrangements that allow for the offset of assets and liabilities in the event of default or early termination of the contract.\nAny amounts of cash collateral received related to these instruments associated with the Company's credit-related contingent features are recorded in Cash and\nequivalents and Accrued liabilities, the latter of which would further offset against the Company's derivative asset balance. Any amounts of cash collateral posted related\nto these instruments associated with the Company's credit-related contingent features are recorded in Prepaid expenses and other current assets, which would further\noffset against the Company's derivative liability balance. Cash collateral received or posted related to the Company's credit-related contingent features is presented in the\nCash provided by operations component of the Consolidated Statements of Cash Flows. The Company does not recognize amounts of non-cash collateral received, such\nas securities, on the Consolidated Balance Sheets. For further information related to credit risk, refer to Note 12 — Risk Management and Derivatives.\n2023 FORM 10-K 68Table of Contents\nThe following tables present information about the Company's derivative assets and liabilities measured at fair value on a recurring basis and indicate the level in the fair\nvalue hierarchy in which the Company classifies the fair value measurement:\nMAY 31, 2023\nDERIVATIVE ASSETS\nDERIVATIVE LIABILITIES"},{"id":"1d742199fb1a86aa8c3f7bcd580d94af","text": ... }

Reviewer: With the uploaded document being about OPEA, output being about Nike fiscals seems unlikely.

Author: @eero-t captured and added OPEA document

```bash
kubectl port-forward svc/chatqna-teirerank 8808:80
```
Follow the below steps in a different terminal.

The TEI Reranking Service reranks the documents returned by the retrieval
service. It consumes the query and list of documents and returns the document
index based on decreasing order of the similarity score. The document
corresponding to the returned index with the highest score is the most relevant
document for the input query.
```
curl http://localhost:8808/rerank \
    -X POST \
    -d '{"query":"What is Deep Learning?", "texts": ["Deep Learning is not...", "Deep learning is..."]}' \
    -H 'Content-Type: application/json'
```

Output is:  `[{"index":1,"score":0.9988041},{"index":0,"score":0.022948774}]`

### TGI Service

Use the following command to forward traffic from your local machine to the service running in the Kubernetes cluster:
```bash
kubectl port-forward svc/chatqna-tgi 9009:80
```
Follow the below steps in a different terminal.

```
curl http://localhost:9009/generate \
  -X POST \
  -d '{"inputs":"What is Deep Learning?","parameters":{"max_new_tokens":17, "do_sample": true}}' \
  -H 'Content-Type: application/json'

```

> TGI service generate text for the input prompt. Here is the expected result from TGI:

Reviewer: ```suggestion
TGI service generates text for the input prompt. Here is the expected result from TGI:
```

```
{"generated_text":"We have all heard the buzzword, but our understanding of it is still growing. It’s a sub-field of Machine Learning, and it’s the cornerstone of today’s Machine Learning breakthroughs.\n\nDeep Learning makes machines act more like humans through their ability to generalize from very large"}
```

> **NOTE**: After launch the TGI, it takes few minutes for TGI server to load LLM model and warm up.

Reviewer: ```suggestion
**NOTE**: After TGI service is started, it takes few minutes to load a LLM model and warm up, before reaching `Ready` state.
```

If you get

```
curl: (7) Failed to connect to localhost port 8008 after 0 ms: Connection refused
```

and the log shows model warm up, please wait for a while and try it later.

```
2024-06-05T05:45:27.707509646Z 2024-06-05T05:45:27.707361Z  WARN text_generation_router: router/src/main.rs:357: `--revision` is not set
2024-06-05T05:45:27.707539740Z 2024-06-05T05:45:27.707379Z  WARN text_generation_router: router/src/main.rs:358: We strongly advise to set it to a known supported commit.
2024-06-05T05:45:27.852525522Z 2024-06-05T05:45:27.852437Z  INFO text_generation_router: router/src/main.rs:379: Serving revision bdd31cf498d13782cc7497cba5896996ce429f91 of model Intel/neural-chat-7b-v3-3
2024-06-05T05:45:27.867833811Z 2024-06-05T05:45:27.867759Z  INFO text_generation_router: router/src/main.rs:221: Warming up model
```

### Dataprep Microservice (Advanced)

Add Knowledge Base via HTTP Links:

```
curl -X POST "http://localhost:6007/v1/dataprep" \
     -H "Content-Type: multipart/form-data" \
     -F 'link_list=["https://opea.dev"]'
```

This command updates a knowledge base by submitting a list of HTTP links for processing.

> Also, you are able to get the file list that you uploaded:

Reviewer: Files could be loaded also by other people:
```suggestion
To get list of uploaded files:
```

```
curl -X POST "http://localhost:6007/v1/dataprep/get_file" \
     -H "Content-Type: application/json"

```

To delete the file/link you uploaded you can use the following commands:

#### Delete link
```
# The dataprep service will add a .txt postfix for link file

> 

Reviewer: Specified `file_path` value already has `.txt` suffix?

Author: @eero-t This is for a url which gets a `.txt`

chatqna-nginx   NodePort   10.201.220.120   <none>        80:30304/TCP   16h
```
We can see that it is serving at port `30304` based on this configuration via a NodePort.

Next step is to get the `<k8s-node-ip-address>` by running:
```bash
kubectl get nodes -o wide
```
The command shows internal IPs for all the nodes in the cluster:
```
NAME       STATUS   ROLES           AGE   VERSION   INTERNAL-IP    EXTERNAL-IP   OS-IMAGE             KERNEL-VERSION       CONTAINER-RUNTIME
minikube   Ready    control-plane   11d   v1.31.0   190.128.49.1   <none>        Ubuntu 22.04.4 LTS   5.15.0-124-generic   docker://27.2.0
```
When using a NodePort, all the nodes in the cluster will be listening at the specified port, which is  `30304` in this example. The `<k8s-node-ip-address>` can be found under INTERNAL-IP. Here it is `190.128.49.1`.

Open a browser to access `http://<k8s-node-ip-address>:${port}`.
From the configuration shown above, it would be `http://190.128.49.1:30304`

Alternatively, You can also choose to use port forwarding as shown previously using:
```bash
kubectl port-forward service/chatqna-nginx 8080:80
```

> and open a browser to access `http://localhost:8080`

Reviewer: Alternative should be easy to see, but this is not due to long text before.  Either move this to beginning of section, or add subheadings for both alternatives.

>  Visit this [link](https://opea-project.github.io/latest/getting-started/README.html#:~:text=tei%2Dembedding%2Dserver%20%20%20%20%20%20%20%20%20%7C-,Interact%20with%20ChatQnA,-%C2%B6) to see how to interact with the UI. 

Reviewer: ```suggestion
 Visit this [link](https://opea-project.github.io/latest/getting-started/README.html#interact-with-chatqna) to see how to interact with the UI. 
```

### Stop the services

> Once you are done with the entire pipeline and wish to stop and remove all the containers, use the command below:

Reviewer: Helm removes everything installed for ChatQnA Helm chart, not just containers.

PS.  Helm does not remove (or update) CRDs added by Helm chart, but happily ChatQnA does not have them currently, so everything **is** removed.

    kubectl get svc
```
   This command will display a list of services along with their network-related details such as cluster IP and ports. 
 ```
NAME                      TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)             AGE
chatqna                   ClusterIP   10.108.186.198   <none>        8888/TCP            8m16s
chatqna-chatqna-ui        ClusterIP   10.102.80.123    <none>        5173/TCP            8m16s
chatqna-data-prep         ClusterIP   10.110.143.212   <none>        6007/TCP            8m16s
chatqna-nginx             NodePort    10.100.224.12    <none>        80:30304/TCP        8m16s
chatqna-redis-vector-db   ClusterIP   10.205.9.19      <none>        6379/TCP,8001/TCP   8m16s
chatqna-retriever-usvc    ClusterIP   10.202.3.15      <none>        7000/TCP            8m16s
chatqna-tei               ClusterIP   10.105.204.12    <none>        80/TCP              8m16s
chatqna-teirerank         ClusterIP   10.115.146.21    <none>        80/TCP              8m16s
chatqna-tgi               ClusterIP   10.108.195.244   <none>        80/TCP              8m16s
kubernetes                ClusterIP   10.92.0.100      <none>        443/TCP             11d
 ```

>    To begin port forwarding, which maps a service's port from the cluster to local host for testing, use:

Reviewer 4: ```suggestion
   To access the services running in your Kubernetes cluster from your local machine, you can set up port forwarding with kubectl.
```

 ```bash
    kubectl port-forward svc/[service-name] [local-port]:[service-port]
   ```
   Replace `[service-name]`, `[local-port]`, and `[service-port]` with the appropriate values from your services list (as shown in the output given by `kubectl get svc`). This setup enables interaction with the microservice directly from the local machine. In another terminal, use `curl` commands to test the functionality and response of the service.

Use `ctrl+c` to end the port-forwarding to test other services.

> ### MegaService Before RAG Dataprep

Reviewer 4: ```suggestion
### Accessing the ChatQnA application
```

This circumvents having to explain what MegaService is, and indeed is the first step before uploading your own documents.

Reviewer 4: Once we have highlighted how to access the different services, using them should be common across the different deployments (on-prem/cloud/docker). Could this not be a point to merge all the variants to a common usage demonstration?

Author: @mkbhanda I did want to do it. Reason I didn't is to include port forward instructions for each microservice, making it easier to quickly copy the command.

Use the following command to forward traffic from your local machine to the service running in the Kubernetes cluster:
```bash
kubectl port-forward svc/chatqna 8888:8888
```
Follow the below steps in a different terminal.

```
curl http://localhost:8888/v1/chatqna -H "Content-Type: application/json" -d '{
     "model": "Intel/neural-chat-7b-v3-3",
     "messages": "What is OPEA?"
     }'

> ```

Reviewer 4: How about adding a note as follows ..

Note: in the curl command, in addition to our prompt, we are specifying  the LLM model to use.

Use the following command to forward traffic from your local machine to the service running in the Kubernetes cluster:
```bash
kubectl port-forward svc/chatqna-data-prep 6007:6007
```
Follow the below steps in a different terminal.

If you want to add/update the default knowledge base, you can use the following
commands. The dataprep microservice extracts the texts from variety of data
sources, chunks the data, embeds each chunk using embedding microservice and
store the embedded vectors in the redis vector database.

this example leverages the OPEA document for its RAG based content. You can download the [OPEA document](https://opea-project.github.io/latest/_downloads/41c91aec1d47f20ca22350daa8c2cadc/what_is_opea.pdf) and upload it using the UI.

Local File `what_is_opea.pdf` Upload:

```
curl -X POST "http://localhost:6007/v1/dataprep" \
     -H "Content-Type: multipart/form-data" \
     -F "files=@./what_is_opea.pdf"
```

This command updates a knowledge base by uploading a local file for processing.
Update the file path according to your environment.

You should see the following output after successful execution:
```
{"status":200,"message":"Data preparation succeeded"}
```

> For advanced usage of the dataprep microservice refer [here](#dataprep-microservice-advanced)

Reviewer 4: "Here" links are not recommended.
```suggestion
Refer to [advanced use of dataprep microservice] (#dataprep-microservice-advanced) to learn more.
```

> ### MegaService After RAG Dataprep

Reviewer 4: ```suggestion
### Accessing ChatQnA application after custom data upload
```

Use the following command to forward traffic from your local machine to the service running in the Kubernetes cluster:
```bash
kubectl port-forward svc/chatqna 8888:8888
```
Similarly, follow the below steps in a different terminal.

```
curl http://localhost:8888/v1/chatqna -H "Content-Type: application/json" -d '{
     "model": "Intel/neural-chat-7b-v3-3",
     "messages": "What is OPEA?"
     }'
```
After uploading the pdf with information about OPEA, we can see that the pdf is being used as a context to answer the question correctly:

```bash
data: b' O'
data: b'PE'
data: b'A'
data: b' ('
data: b'Open'
data: b' Platform'
data: b' for'
data: b' Enterprise'
data: b' AI'
data: b')',
.
.
.
data: b' systems'
data: b'.'
data: b''
data: b''
data: [DONE]
```

The above output has been parsed into the below sentence which shows how the LLM has picked up the right context to answer the question correctly after the document upload:
```

> OPEN Platform for Enterprise AI (Open Platform for Enterprise AI) is a framework that focuses on creating and evaluating open, multi-provider, robust, and composable generative AI (GenAI) solutions. It aims to facilitate the implementation of enterprise-grade composite GenAI solutions, particularly Retrieval Augmented Generative AI (RAG), by simplifying the integration of secure, performant, and cost-effective GenAI workflows into business systems.

Reviewer 4: ```suggestion
Open Platform for Enterprise AI (Open Platform for Enterprise AI) is a framework that focuses on creating and evaluating open, multi-provider, robust, and composable generative AI (GenAI) solutions. It aims to facilitate the implementation of enterprise-grade composite GenAI solutions, particularly Retrieval Augmented Generative AI (RAG), by simplifying the integration of secure, performant, and cost-effective GenAI workflows into business systems.
```

```

### TEI Embedding Service

> Use the following command to forward traffic from your local machine to the service running in the Kubernetes cluster:

Reviewer 4: ```suggestion
Use the following command to forward traffic from your local machine to the TEI service running in your Kubernetes cluster:
```

```bash
kubectl port-forward svc/chatqna-tei 6006:80
```
Follow the below steps in a different terminal.

The TEI embedding service takes in a string as input, embeds the string into a
vector of a specific length determined by the embedding model and returns this
embedded vector.

```
curl http://localhost:6006/embed \
    -X POST \
    -d '{"inputs":"What is Deep Learning?"}' \
    -H 'Content-Type: application/json'
```

In this example the embedding model used is "BAAI/bge-base-en-v1.5", which has a vector size of 768. So the output of the `curl` command is a embedded vector of
length 768.

### Retriever Microservice

> Use the following command to forward traffic from your local machine to the service running in the Kubernetes cluster:

Reviewer 4: ```suggestion
Use the following command to forward traffic from your local machine to the Retriever service running in your Kubernetes cluster:
```

```bash
kubectl port-forward svc/chatqna-retriever-usvc 7000:7000
```
Follow the below steps in a different terminal.

To consume the retriever microservice, you need to generate a mock embedding
vector by Python script. The length of embedding vector is determined by the
embedding model. Here we use the

> model EMBEDDING_MODEL_ID="BAAI/bge-base-en-v1.5", which vector size is 768.

Reviewer 4: ```suggestion
model EMBEDDING_MODEL_ID="BAAI/bge-base-en-v1.5", which creates a vector of size 768.
```

Check the vector dimension of your embedding model and set
`your_embedding` dimension equal to it.

```
export your_embedding=$(python3 -c "import random; embedding = [random.uniform(-1, 1) for _ in range(768)]; print(embedding)")

curl http://localhost:7000/v1/retrieval \
  -X POST \
  -d "{\"text\":\"test\",\"embedding\":${your_embedding}}" \
  -H 'Content-Type: application/json'
```

> The output of the retriever microservice comprises of the a unique id for the

Reviewer 4: ```suggestion
The output of the retriever microservice comprises of a unique id for the
```

request, initial query or the input to the retrieval microservice, a list of top
`n` retrieved documents relevant to the input query, and top_n where n refers to
the number of documents to be returned.

> The output is retrieved text that relevant to the input data:

Reviewer 4: ```suggestion
The output is retrieved text that is relevant to the input data:
```

```
### TEI Reranking Service

> Use the following command to forward traffic from your local machine to the service running in the Kubernetes cluster:

Reviewer 4: ```suggestion
Use the following command to forward traffic from your local machine to the Reranking service running in the Kubernetes cluster:
```

```bash
kubectl port-forward svc/chatqna-teirerank 8808:80
```
Follow the below steps in a different terminal.

The TEI Reranking Service reranks the documents returned by the retrieval
service. It consumes the query and list of documents and returns the document

> index based on decreasing order of the similarity score. The document

Reviewer 4: ```suggestion
indices based on decreasing order of the similarity score. The document
```

corresponding to the returned index with the highest score is the most relevant
document for the input query.
```
curl http://localhost:8808/rerank \
    -X POST \
    -d '{"query":"What is Deep Learning?", "texts": ["Deep Learning is not...", "Deep learning is..."]}' \
    -H 'Content-Type: application/json'
```

Output is:  `[{"index":1,"score":0.9988041},{"index":0,"score":0.022948774}]`

### TGI Service

> Use the following command to forward traffic from your local machine to the service running in the Kubernetes cluster:

Reviewer 4: ```suggestion
Use the following command to forward traffic from your local machine to the TGI service running in your Kubernetes cluster:
```

```bash
kubectl port-forward svc/chatqna-tgi 9009:80
```
Follow the below steps in a different terminal.

```
curl http://localhost:9009/generate \
  -X POST \
  -d '{"inputs":"What is Deep Learning?","parameters":{"max_new_tokens":17, "do_sample": true}}' \
  -H 'Content-Type: application/json'
```

TGI service generates text for the input prompt. Here is the expected result from TGI:

```
{"generated_text":"We have all heard the buzzword, but our understanding of it is still growing. It’s a sub-field of Machine Learning, and it’s the cornerstone of today’s Machine Learning breakthroughs.\n\nDeep Learning makes machines act more like humans through their ability to generalize from very large"}
```

> **NOTE**: After TGI service is started, it takes few minutes to load a LLM model and warm up, before reaching `Ready` state.

Reviewer 4: ```suggestion
**NOTE**: After TGI service is started, it takes a few minutes, to load the LLM model and warm up, before it reaches the `Ready` state.
```

If you get

```
curl: (7) Failed to connect to localhost port 8008 after 0 ms: Connection refused
```

> and the log shows model warm up, please wait for a while and try it later.

Reviewer 4: ```suggestion
and the log shows model warm up, please wait for a while and retry.
```

```
2024-06-05T05:45:27.707509646Z 2024-06-05T05:45:27.707361Z  WARN text_generation_router: router/src/main.rs:357: `--revision` is not set
2024-06-05T05:45:27.707539740Z 2024-06-05T05:45:27.707379Z  WARN text_generation_router: router/src/main.rs:358: We strongly advise to set it to a known supported commit.
2024-06-05T05:45:27.852525522Z 2024-06-05T05:45:27.852437Z  INFO text_generation_router: router/src/main.rs:379: Serving revision bdd31cf498d13782cc7497cba5896996ce429f91 of model Intel/neural-chat-7b-v3-3
2024-06-05T05:45:27.867833811Z 2024-06-05T05:45:27.867759Z  INFO text_generation_router: router/src/main.rs:221: Warming up model
```

### Dataprep Microservice (Advanced)

Add Knowledge Base via HTTP Links:

> 

Reviewer 4: Please add 

Once you have set up port forward for the dataprep service, you can upload, delete, and list documents.

# Multi-node on-prem deployment with  TGI on Xeon Scalable processors on a K8s cluster using Helm Charts

This deployment section covers multi-node on-prem deployment of the ChatQnA
example with OPEA comps to deploy using the TGI service. There are several
slice-n-dice ways to enable RAG with vectordb and LLM models, but here we will
be covering one option of doing it for convenience : we will be showcasing  how
to build an e2e chatQnA with Redis VectorDB and neural-chat-7b-v3-3 model,
deployed on a Kubernetes cluster. For more information on how to setup a Xeon based Kubernetes cluster along with the development pre-requisites,
please follow the instructions here (*** ### Kubernetes Cluster and Development Environment***). 

> For a quick introduction on Helm Charts, visit the helm section in  (**getting started**)

Reviewer 5: The link to your getting started doc (part of this PR) would be:
```
[Getting Started with Kubernetes](./k8s_getting_started.md)
```

## Overview

There are several ways to setup a ChatQnA use case. Here in this tutorial, we
will walk through how to enable the below list of microservices from OPEA
GenAIComps to deploy a multi-node TGI megaservice solution.

> > **Note:** ChatQnA can also be deployed on a single node using Kubernetes, provided that all pods are configured to run on the same node.

Reviewer 5: I'd move this note after the list of microservices so it flows better to the reader.

> 1. Data Prep

Reviewer 5: Numbered lists imply a sequential order.  This feels more like just a list, so use a bullet (unordered) list instead;
```
- Data Prep
- Embedding
- Retriever
- Reranking
- LLM with TGI
```

Author: @dbkinder I kept this similar to xeon.md

2. Embedding
3. Retriever
4. Reranking
5. LLM with TGI

## Prerequisites

### Install Helm
First, ensure that Helm (version >= 3.15) is installed on your system. Helm is an essential tool for managing Kubernetes applications. It simplifies the deployment and management of Kubernetes applications using Helm charts. 

> For detailed installation instructions, please refer to the [Helm Installation Guide](https://helm.sh/docs/intro/install/)

Reviewer 5: delete "please"

### Clone Repository 

> First step is to clone the GenAIInfra which is the containerization and cloud native suite for OPEA, including artifacts to deploy ChatQnA in a cloud native way.

Reviewer 5: We already had what was said to be a First in the step above, so maybe "Next step..."

```bash

> git clone https://github.com/opea-project/GenAIInfra.git

Reviewer 5: What directory do you want them to be in before doing this clone?

Author: This can be done anywhere. It's self-contained to be able to run chatqna via helm.

```
Checkout the release tag
```
cd GenAIInfra/helm-charts/
git checkout tags/v1.0
```
### HF Token
The example can utilize model weights from HuggingFace and langchain.

Setup your [HuggingFace](https://huggingface.co/) account and generate
[user access token](https://huggingface.co/docs/transformers.js/en/guides/private#step-1-generating-a-user-access-token).

Setup the HuggingFace token
```
export HF_TOKEN="Your_Huggingface_API_Token"
```

### Proxy Settings

> Make sure to setup Proxies if you are behind a firewall.

Reviewer 5: I'd delete this sentence since it's covered below.

For services requiring internet access, such as the LLM microservice, embedding service, reranking service, and other backend services, proxy settings can be essential. These settings ensure services can download necessary content from the internet, especially when behind a corporate firewall.
Proxy can be set in the `values.yaml` file, like so:
Open the `values.yaml` file using an editor
```bash

> vi GenAIInfra/helm-charts/chatqna/values.yaml

Reviewer 5: In the previous step we did a cd into the GenAIInfra/helm-charts directory so this command would fail.

data: b'2'
data: b'0'
data: b'2'
data: b'3'
data: b','
data: b' N'
data: b'I'
data: b'KE'
data: b','
data: b' Inc'
data: b'.'
data: b' achieved'
data: b' record'
data: b' Rev'
data: b'en'
data: b'ues'
data: b' of'
data: b' $'
data: b'5'
data: b'1'
data: b'.'
data: b'2'
data: b' billion'
data: b'.'
data: b'</s>'
data: [DONE]
```
## Launch UI
### Basic UI
To access the frontend, open the following URL in your browser: 
`http://{k8s-node-ip-address}:${port}`
You can find the NGINX port using the following command:
```bash
export port=$(kubectl get service chatqna-nginx --output='jsonpath={.spec.ports[0].nodePort}')
echo $port
```
Open a browser to access `http://<k8s-node-ip-address>:${port}`
 
 By default, the UI runs on port 5173 internally. If you prefer to use a different host port to access the frontend, you can modify the port mapping in the `GenAIInfra/helm-charts/chatqna/values.yaml` file as shown below:
```
chatqna-ui:
   image:
     repository: "opea/chatqna-ui"
     tag: "latest"
   containerPort: "5173"
```
### Stop the services
Once you are done with the entire pipeline and wish to stop and remove all the containers, use the command below:
```
helm uninstall chatqna

> ```

Reviewer 5: missing a newline character on the last line

This use case employs a tailored combination of Helm charts and `values.yaml` configurations to deploy the following components and tools:
|use case components | Tools |   Model     | Service Type |
|----------------     |--------------|-----------------------------|-------|
|Data Prep            |  LangChain   | NA                       |OPEA Microservice |
|VectorDB             |  Redis       | NA                       |Open source service|
|Embedding            |   TEI        | BAAI/bge-base-en-v1.5    |OPEA Microservice |
|Reranking            |   TEI        | BAAI/bge-reranker-base   | OPEA Microservice |
|LLM                  |   TGI        |Intel/neural-chat-7b-v3-3 |OPEA Microservice |
|UI                   |              | NA                       | Gateway Service |

> Tools and models mentioned in the table are configurable either through the

Reviewer 5: Add a blank line after a table. Otherwise, this text will be included in the table, as shown if you viewed this file:

![image](https://github.com/user-attachments/assets/fd807dbc-cdbf-4603-9213-81676256e925)

# Multi-node on-prem deployment with  TGI on Xeon Scalable processors on a K8s cluster using Helm Charts

This deployment section covers multi-node on-prem deployment of the ChatQnA
example with OPEA comps to deploy using the TGI service. There are several
slice-n-dice ways to enable RAG with vectordb and LLM models, but here we will
be covering one option of doing it for convenience : we will be showcasing  how
to build an e2e chatQnA with Redis VectorDB and neural-chat-7b-v3-3 model,
deployed on a Kubernetes cluster. For more information on how to setup a Xeon based Kubernetes cluster along with the development pre-requisites,
please follow the instructions here (*** ### Kubernetes Cluster and Development Environment***). 

> For a quick introduction on Helm Charts, visit the helm section in  [Getting Started with Kubernetes for ChatQnA](./k8s_getting_started.md)

Author: @tylertitsworth addressed line breaks
The text itself is from a template, I will make sure to bring this up to edit this across all templates

Tools and models mentioned in the table are configurable either through the
environment variable or `values.yaml` 

Set a new [namespace](#create-and-set-namespace) and switch to it if needed

To enable UI, uncomment the lines `54-58` in `GenAIInfra/helm-charts/chatqna/values.yaml`:
```bash
chatqna-ui:
   image:
     repository: "opea/chatqna-ui"
     tag: "latest"
   containerPort: "5173"
```

Next, we will update the dependencies for all Helm charts in the specified directory and ensure the `chatqna` Helm chart is ready for deployment by updating its dependencies as defined in the `Chart.yaml` file.

```bash
# all Helm charts in the specified directory have their 
# dependencies up-to-date, facilitating consistent deployments.
./update_dependency.sh

# "chatqna" here refers to the directory name that contains the Helm 
# chart for the ChatQnA application 
helm dependency update chatqna
```

Set the necessary environment variables to setup the use case
```bash
export MODELDIR="/mnt/opea-models"  #export MODELDIR="null" if you don't want to cache the model.  
export MODELNAME="Intel/neural-chat-7b-v3-3"
export EMBEDDING_MODELNAME="BAAI/bge-base-en-v1.5"

> export RERANKER_MODELNAME="BAAI/bge-reranker-base"

Author: @tylertitsworth As discussed, will open a github feature request and add you in it
(since it's currently spread across multiple charts)

```

## Deploy the use case
In this tutorial, we will be deploying using Helm with the provided chart. The Helm install commands will initiate all the aforementioned services as Kubernetes pods.

```bash
helm install chatqna chatqna \
    --set global.HUGGINGFACEHUB_API_TOKEN=${HF_TOKEN} \
    --set global.modelUseHostPath=${MODELDIR} \
    --set tgi.LLM_MODEL_ID=${MODELNAME} \
    --set tei.EMBEDDING_MODEL_ID=${EMBEDDING_MODELNAME} \
    --set teirerank.RERANK_MODEL_ID=${RERANKER_MODELNAME}
```

**OUTPUT:**
```bash
NAME: chatqna
LAST DEPLOYED: Thu Sep  5 13:40:20 2024
NAMESPACE: chatqa
STATUS: deployed
REVISION: 1

> 

Author: @tylertitsworth The installation notes are under Validate microservice --> Check the pod status which are the following sections
Do you have any suggestions to make it better? or add something?

```
> **Note:** Use `kubectl get pods -o wide` to check the nodes that the respective pods are running on

When issues are encountered with a pod in the Kubernetes deployment, there are two primary commands to diagnose and potentially resolve problems:
1. **Checking Logs**: To view the logs of a specific pod, which can provide insight into what the application is doing and any errors it might be encountering, use:
    ```bash
    kubectl logs [pod-name]
    ```
2. **Describing Pods**: For a detailed view of the pod's current state, its configuration, and its operational events, run:
	```bash
    kubectl describe pod [pod-name]
    ```
For example, if the status of the TGI service does not show 'Running', describe the pod using the name from the above table:
```bash
kubectl describe pod chatqna-tgi-778bb6598f-cv5cg
```
or check logs using:
```bash
kubectl logs chatqna-tgi-778bb6598f-cv5cg
```

## Interacting with ChatQnA deployment
This section will walk you through what are the different ways to interact with
the microservices deployed

Before starting the validation of microservices, check the network configuration of services using:

> ```bash
>     kubectl get svc
>    ```

Reviewer 6: ```suggestion
```bash
kubectl get svc
```
```

input parameters, it takes in a string, embeds it into a vector using the TEI
embedding service and pads other default parameters that are required for the
retrieval microservice and returns it.
```
curl http://localhost:6000/v1/embeddings\
  -X POST \
  -d '{"text":"hello"}' \
  -H 'Content-Type: application/json'
```

### Retriever Microservice
Use the following command to forward traffic from your local machine to the service running in the Kubernetes cluster:
```bash
kubectl port-forward svc/chatqna-retriever-usvc 7000:7000
```
Follow the below steps in a different terminal.

To consume the retriever microservice, you need to generate a mock embedding
vector by Python script. The length of embedding vector is determined by the
embedding model. Here we use the
model EMBEDDING_MODEL_ID="BAAI/bge-base-en-v1.5", which vector size is 768.

Check the vector dimension of your embedding model and set
`your_embedding` dimension equal to it.

```
export your_embedding=$(python3 -c "import random; embedding = [random.uniform(-1, 1) for _ in range(768)]; print(embedding)")

curl http://localhost:7000/v1/retrieval \
  -X POST \
  -d "{\"text\":\"test\",\"embedding\":${your_embedding}}" \

>   -H 'Content-Type: application/json'

Reviewer 6: You could probably add the use of `jq` here to truncate just the text output. So it's more readable.

export RERANKER_MODELNAME="BAAI/bge-reranker-base"
```

> **Note:**
> 
> Setting `MODELDIR` to an empty string will download the models without sharing them among worker nodes. This configuration is intended as a quick setup for testing in a single-node environment.
> 
> In a multi-node environment, go to every K8s worker node to make sure that a ${MODELDIR} directory exists and is writable.
> 

> > Another option is to to use K8s persistent volume to share the model data files. For more information click [here](https://github.com/opea-project/GenAIInfra/blob/main/helm-charts/README.md#using-persistent-volume).

Reviewer: ```suggestion
> Another option is to to use k8s persistent volume to share the model data files. For more information see [Using Persistent Volume](https://github.com/opea-project/GenAIInfra/blob/main/helm-charts/README.md#using-persistent-volume).
```
